### PR TITLE
feature benchmark: add scenarios ManySmallInserts and ManySmallUpdates

### PR DESCRIPTION
**Pull request by nrainer-materialize**
Nightly: https://buildkite.com/materialize/nightly/builds/7305

### CODE REVIEW

**Comment by nrainer-materialize** [comment]
Hey @danhhz,
Are these scenarios that should profit from persist inline writes? If not, what would I need to change?

**Comment by danhhz** [comment]
This just gets run single-threaded, yeah? If so, I think the inserts one should be pretty sensitive to inline writes. That many small inserts in a row should hopefully even exercise the backpressure to keep persist state size from blowing up.

Less clear on the updates because (1) there's quite a bit of overhead to read-write statements in mz and (2) it looks like it might be updating 10% of the data on each statement, which I'd guess would be well above the threshold for inline writes.

File: misc/python/materialize/feature_benchmark/scenarios/benchmark_main.py [hunk 1]
```diff
@@ -227,6 +227,32 @@ def benchmark(self) -> MeasurementSource:
...
+class ManySmallInserts(DML):
+    """Measure the time it takes for several small INSERT statements to return."""
+
+    def init(self) -> Action:
+        return self.table_ten()
+
+    def benchmark(self) -> MeasurementSource:
+        statements = []
+        for _ in range(0, 10000):
+            statements.append(f"> INSERT INTO t1 VALUES ({random.randint(0, 100000)})")
```
Comment (def-) [inline misc/python/materialize/feature_benchmark/scenarios/benchmark_main.py:239]: We should allow having a fixed seed, as in platform-checks for example.

**Review by def- — CHANGES_REQUESTED** [review]

**Comment by nrainer-materialize** [comment]
> Less clear on the updates because (1) there's quite a bit of overhead to read-write statements in mz and (2) it looks like it might be updating 10% of the data on each statement, which I'd guess would be well above the threshold for inline writes.

Well, the table contains only ten rows, so 10% would be a single row. 😀 
Should I remove the where condition to update all rows to avoid read operations? What role does the index play?

**Comment by danhhz** [comment]
> Well, the table contains only ten rows, so 10% would be a single row. 😀

Oh, ha! I read that wrong. Yeah that seems fine then.

> Should I remove the where condition to update all rows to avoid read operations? What role does the index play?

Any update is a read then write in materialize because it has to find the old value to issue the -1 retraction. The index should speed up finding this old value, so I'd expect it to help.

**Review by def- — APPROVED** [review]

**Comment by nrainer-materialize** [comment]
> Updates certainly seems within the realm of the speedup I'd expect.

Great 👍 

> I'm surprised at the relative time between the two. My (probably incorrect) read of the PR is that we do 10,000 of each and inserts are definitely way faster than updates in mz. Is this measured in seconds? The BOE math for that doesn't check out.

Yes, wallclock is measured in seconds. Why is it surprising that inserts are faster than updates? Didn't you say that each update requires a read and a retraction in addition to the write of the new value?

> Any idea what blob (S3? Files?) and consensus (CRDB with fsync? Our in-mem CRDB docker image?) this is using?

Not sure if I can fully answer your question.

What I can tell is that we are running Cockroach (`v23.1.11`) in Docker with `--start-single-node` and these configuration properties:
```
    "COCKROACH_ENGINE_MAX_SYNC_DURATION_DEFAULT=120s",
    "COCKROACH_LOG_MAX_SYNC_DURATION=120s",
```

Cockroach is NOT running in memory.

The feature benchmark setup is specified in `SERVICES` in `test/feature-benchmark/mzcompose.py`. We use the mz configuration specified in `DEFAULT_SYSTEM_PARAMETERS` in `misc/python/materialize/mzcompose/__init__.py`.

**Pull request merged**